### PR TITLE
MuTau Pre-Selection

### PR DESCRIPTION
**Pull request by lucasrussell01**
Modified event selection to generate event Tuples comparable to those used in HIG-21-001
- Select most isolated ref_muon or tau candidate in event more than one pass selection
- Extra lepton veto variables set to true if extra muon, extra electron or a muon pair are present.
- Extra lepton variables introduced in TauTuple

### CODE REVIEW

File: Production/interface/Selectors.h [hunk 1]
```diff
@@ -15,8 +15,17 @@ struct TagObject {
...
+    bool extramuon;
+    bool extraelectron;
+    bool dimuon;
 };
 
+bool muonveto (const std::vector<pat::Muon>& muons, const pat::Muon *ref_muon, const reco::Vertex& primaryVertex);
+
+bool electronveto (const std::vector<pat::Electron>& electrons, const reco::Vertex& primaryVertex, const float rho);
+
+bool dimuonveto (const std::vector<pat::Muon>& muons, const pat::Muon *ref_muon, const reco::Vertex& primaryVertex);
```
Comment (kandrosov) [inline Production/interface/Selectors.h:27]: - `ref_muon` is not needed for dimionveto
- better to rename functions to `hasExtraMuon` etc.
- define functions directly in an anonymous namespace in Selectors.cc

File: Production/interface/Selectors.h [hunk 2]
```diff
@@ -26,7 +35,7 @@ struct TauJetSelector {
                           const std::vector<pat::Muon>& muons, const pat::MET& met,
                           const reco::Vertex& primaryVertex,
                           const pat::TriggerObjectStandAloneCollection& triggerObjects,
-                          const edm::TriggerResults& triggerResults);
+                          const edm::TriggerResults& triggerResults, const float rho);
```
Comment (kandrosov) [inline Production/interface/Selectors.h:38]: `const float rho` -> `float rho`

File: Production/interface/TauAnalysis.h [hunk 3]
```diff
@@ -47,6 +48,7 @@ double CalculateDeltaEtaCrack(double eta);
 // Based on https://github.com/cms-sw/cmssw/blob/CMSSW_9_4_X/RecoTauTag/RecoTau/plugins/PATTauDiscriminationByMVAIsolationRun2.cc#L218
 bool CalculateGottfriedJacksonAngleDifference(const pat::Tau& tau, double& gj_diff);
 
-double PFIsolation(const pat::Muon& muon);
+double PFRelIsolation(const pat::Muon& muon);
 
+double PFRelIsolation_e(const pat::Electron& electron, const float rho);
```
Comment (kandrosov) [inline Production/interface/TauAnalysis.h:53]: - `PFRelIsolation_e` -> `PFRelIsolation`
- `const float rho` -> `float rho`

File: Production/plugins/TauTupleProducer.cc [hunk 4]
```diff
@@ -271,7 +271,7 @@ class TauTupleProducer : public edm::EDAnalyzer {
                               *isoTracks, *lostTracks, genParticles, genJets, requireGenMatch,
                               requireGenORRecoTauMatch, applyRecoPtSieve);
         const auto [tauJets, tagObj] = selector->Select(event, builder.GetTauJets(), *electrons, *muons,
-                                                           METs->at(0), PV, *triggerObjects, *triggerResults);
+                                                           METs->at(0), PV, *triggerObjects, *triggerResults, static_cast<float>(*rho));
```
Comment (kandrosov) [inline Production/plugins/TauTupleProducer.cc:274]: is static_cast necessary?

File: Production/src/Selectors.cc [hunk 5]
```diff
@@ -15,12 +15,47 @@ std::shared_ptr<TauJetSelector> TauJetSelector::Make(const std::string& name)
     throw analysis::exception("Unknown selector name = '%1%'") % name;
 }
 
+bool muonveto (const std::vector<pat::Muon>& muons, const pat::Muon *ref_muon, const reco::Vertex& primaryVertex){
+    for(const pat::Muon& muon : muons) {
+        if(muon.pt() > 10 && std::abs(muon.eta()) < 2.4 && muon.isMediumMuon() && PFRelIsolation(muon) < 0.30 && std::abs(muon.muonBestTrack()->dxy(primaryVertex.position())) < 0.2
+                && std::abs(muon.muonBestTrack()->dz(primaryVertex.position())) < 0.0045&& &muon != ref_muon){
```
Comment (kandrosov) [inline Production/src/Selectors.cc:21]: Looks like dxy and dz cuts are flipped

File: Production/src/Selectors.cc [hunk 6]
```diff
@@ -15,12 +15,47 @@ std::shared_ptr<TauJetSelector> TauJetSelector::Make(const std::string& name)
...
+            return true;
+        }
+    }
+    return false;
+}
+
+bool electronveto (const std::vector<pat::Electron>& electrons, const reco::Vertex& primaryVertex, const float rho){
```
Comment (kandrosov) [inline Production/src/Selectors.cc:28]: `primaryVertex` is not used, remove it from the list of arguments.

File: Production/src/Selectors.cc [hunk 7]
```diff
@@ -15,12 +15,47 @@ std::shared_ptr<TauJetSelector> TauJetSelector::Make(const std::string& name)
...
+    for(const pat::Electron& electron : electrons) {
+        if(electron.pt() > 10 && std::abs(electron.eta()) < 2.5 && electron.electronID("mvaEleID-Fall17-noIso-V2-wp90") > 0.5f && PFRelIsolation_e(electron, rho)<0.3){
+            return true;
+        }
+    }
+    return false;
+}
+
+bool dimuonveto (const std::vector<pat::Muon>& muons, const pat::Muon *ref_muon, const reco::Vertex& primaryVertex){
+    std::vector<pat::Muon> dimuon_candidates; // vector of all muons that pass selection
```
Comment (kandrosov) [inline Production/src/Selectors.cc:38]: - should be `std::vector<const pat::Muon*>`

File: Production/src/Selectors.cc [hunk 8]
```diff
@@ -15,12 +15,47 @@ std::shared_ptr<TauJetSelector> TauJetSelector::Make(const std::string& name)
...
+    for(const pat::Muon& muon : muons) {
+        if(muon.pt() > 15 && std::abs(muon.eta()) < 2.4 && muon.isLooseMuon() && PFRelIsolation(muon) < 0.30 && std::abs(muon.muonBestTrack()->dxy(primaryVertex.position())) < 0.2
+                && std::abs(muon.muonBestTrack()->dz(primaryVertex.position())) < 0.0045)
```
Comment (kandrosov) [inline Production/src/Selectors.cc:41]: Looks like dxy and dz cuts are flipped

File: Production/src/Selectors.cc [hunk 9]
```diff
@@ -15,12 +15,47 @@ std::shared_ptr<TauJetSelector> TauJetSelector::Make(const std::string& name)
...
+                    dimuon_candidates.push_back(muon);
+    }
+    for (const pat::Muon& muon1 : dimuon_candidates) { // look at all possible matches
+        for (const pat::Muon& muon2 : dimuon_candidates) {
```
Comment (kandrosov) [inline Production/src/Selectors.cc:45]: better:
```c++
for(size_t m1 = 0; m1 < dimoun_cands.size(); ++m1) {
  for(size_t m2 = m1 + 1; m2 < dimuon_cands.size(); ++m2) {
    ...
  }
}
```

File: Production/src/Selectors.cc [hunk 10]
```diff
@@ -33,21 +68,25 @@ TauJetSelector::Result MuTau::Select(const edm::Event& event, const std::deque<T
...
-        if(!(muon.pt() > 28 && std::abs(muon.eta()) < 2.1 && muon.isMediumMuon() && PFIsolation(muon) < 0.15
+        if(!(muon.pt() > 25 && std::abs(muon.eta()) < 2.1 && muon.isMediumMuon() && PFRelIsolation(muon) < 0.15
                 && std::abs(muon.muonBestTrack()->dxy(primaryVertex.position())) < 0.2
                 && std::abs(muon.muonBestTrack()->dz(primaryVertex.position())) < 0.0045))
             continue;
-        if(!ref_muon || ref_muon->pt() < muon.pt())
+        if(!ref_muon || PFRelIsolation(*ref_muon) < PFRelIsolation(muon) || (PFRelIsolation(*ref_muon) == PFRelIsolation(muon) && ref_muon->pt() < muon.pt())){
             ref_muon = &muon;
-
+            // std::cout<<"Reference muon assigned" << std::endl;
```
Comment (kandrosov) [inline Production/src/Selectors.cc:86]: remove std::couts

File: Production/src/Selectors.cc [hunk 11]
```diff
@@ -33,21 +68,25 @@ TauJetSelector::Result MuTau::Select(const edm::Event& event, const std::deque<T
...
+
     const pat::Muon *ref_muon = nullptr;
     for(const pat::Muon& muon : muons) {
-        if(!(muon.pt() > 28 && std::abs(muon.eta()) < 2.1 && muon.isMediumMuon() && PFIsolation(muon) < 0.15
+        if(!(muon.pt() > 25 && std::abs(muon.eta()) < 2.1 && muon.isMediumMuon() && PFRelIsolation(muon) < 0.15
                 && std::abs(muon.muonBestTrack()->dxy(primaryVertex.position())) < 0.2
                 && std::abs(muon.muonBestTrack()->dz(primaryVertex.position())) < 0.0045))
             continue;
-        if(!ref_muon || ref_muon->pt() < muon.pt())
+        if(!ref_muon || PFRelIsolation(*ref_muon) < PFRelIsolation(muon) || (PFRelIsolation(*ref_muon) == PFRelIsolation(muon) && ref_muon->pt() < muon.pt())){
```
Comment (kandrosov) [inline Production/src/Selectors.cc:84]: PFRelIsolation a < PFRelIsolation b means that a is more isolated than b. The condition should be inverted.

File: Production/src/TauAnalysis.cc [hunk 12]
```diff
@@ -48,4 +48,19 @@ double PFIsolation(const pat::Muon& muon)
     return abs_iso / muon.pt();
 }
 
+double PFRelIsolation_e(const pat::Electron& electron, const float rho)
+{   
+    const std::vector<float> eta_lower = {0.0, 1.0, 1.479, 2.0, 2.2, 2.3, 2.4};
+    const std::vector<float> ea_values = {0.1440, 0.1562, 0.1032, 0.0859, 0.1116, 0.1321, 0.1654};
+
+    auto eta_bin = std::lower_bound(eta_lower.begin(), eta_lower.end(), electron.eta()); // iterator pointing to corresponding eta value
```
Comment (kandrosov) [inline Production/src/TauAnalysis.cc:56]: std::abs(eta)

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: Production/src/Selectors.cc (hunk 7 above)
Comment (lucasrussell01) [reply Production/src/Selectors.cc:38]: in both lines 37 and 38?

**Review by kandrosov — APPROVED** [review]

**Pull request merged**